feat(client): Implement read-only `StorageState` functionality on `ClientAdapter`

haikoschol opened the pull request:
## Changes

This PR should cover all changes mentioned in [the issue](https://github.com/ChainSafe/gossamer/issues/4467), except for changing `StorageState.Entries()` to return a `statemachine.PairsIter`. As we discussed on Slack, this will be postponed since it is only used in RPC, which is currently not a priority.

Regarding `LoadCode()`; The implementation on `ClientAdapter` expects a block hash as the argument. [The current implementation in `state.InMemoryStorageState`](https://github.com/ChainSafe/gossamer/blob/dad5d11e83dcaf7d1f308365929f7415c696bf78/dot/state/inmemory_storage.go#L278) calls `GetStorage()`, which expects a state root hash. I'm pretty sure this is a long existing bug that hasn't been noticed because `LoadCode()` is called with `nil` on startup, which defaults to the best block.

## Tests

```sh
go test github.com/ChainSafe/gossamer/internal/client/adapter
```

## Issues

#4467

haikoschol committed:
define StorageProvider interface

haikoschol committed:
implement StorageProvider on ClientAdapter

haikoschol committed:
implement read-only StorageState methods on ClientAdapter

haikoschol committed:
lint

haikoschol committed:
Merge branch 'development' into haiko/read-only-storage-state

dimartiro commented on internal/client/adapter/client_adapter.go:
```diff
@@ -398,8 +443,10 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) SetHighestRoundAndSetID(round,
+// GetRoundAndSetID returns the finalised round and setID
 func (ca *ClientAdapter[H, Hasher, N, E, Header]) GetRoundAndSetID() (uint64, uint64) {
-	panic("unimplemented")
+	round, setID, _ := ca.GetHighestRoundAndSetID() // TODO check this is correct
```
I think it's correct, you can remove the TODO

dimartiro commented on internal/client/adapter/client_adapter.go:
```diff
@@ -495,53 +542,204 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) StoreTrie(*rtstorage.TrieState
 	panic("unimplemented")
 }
 
+// GetStateRootFromBlock returns the state root of the block with the given hash.
```
You can add the case when it's call with nil to the doc

dimartiro commented on internal/client/adapter/client_adapter.go:
```diff
@@ -495,53 +542,204 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) StoreTrie(*rtstorage.TrieState
+// LoadCode returns the runtime blob for the given block hash.
 func (ca *ClientAdapter[H, Hasher, N, E, Header]) LoadCode(hash *common.Hash) ([]byte, error) {
-	panic("unimplemented")
+	// InmemoryStorageState.LoadCode() calls GetStorage() but I'm pretty sure `hash` is meant to be a block hash,
```
I'm not sure if I understand this comment. Hash is meant to be the block hash, but `InmemoryStorageState .GetStorage` is also using the block hash, or do I miss something?

haikoschol replied:
`InmemoryStorageState.GetStorage()` [expects a "storage hash"](https://github.com/ChainSafe/gossamer/blob/205b72beca190262c12bd13ccd38d12bd026d06f/dot/state/inmemory_storage.go#L179), i.e. state root hash. It uses a `state.Tries` object to access storage, which maintains [a mapping from root hash to `Trie`](https://github.com/ChainSafe/gossamer/blob/476115aa6f5c0ebe70b7217ec4f15eb19f38673a/dot/state/inmemory_tries.go#L37).

dimartiro replied:
oh got it! anyways we should find a way to solve this 🤔

haikoschol replied:
solved in #4792

dimartiro commented on internal/client/adapter/client_adapter.go:
```diff
@@ -495,53 +542,204 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) StoreTrie(*rtstorage.TrieState
+		return common.Hash{}, err
+	}
+
+	return common.Blake2bHash(code)
```
I think you need to use the generic Hasher here and then convert that result into `common.Hash`

dimartiro commented on internal/client/adapter/client_adapter.go:
```diff
@@ -556,6 +754,138 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) SetBlockTree(blocktree *blockt
+	return stateAt.Keys(statemachine.IterArgs{
+		Prefix:           prefix,
+		StartAt:          startKey,
+		StartAtExclusive: false, // TODO: check that this is correct
```
You can check this by using the substrate tests (if any)

haikoschol replied:
Didn't even need tests, I could have just looked at the Rust code. 🤦 

[`Client::storage_keys()` calls `KeysIter::new()`](https://github.com/paritytech/polkadot-sdk/blob/9991b1a804202bfa0beee1bdd7f78e0d1fd642a3/substrate/client/service/src/client/client.rs#L1421) which [sets `start_at_exclusive` to `true`](https://github.com/paritytech/polkadot-sdk/blob/9991b1a804202bfa0beee1bdd7f78e0d1fd642a3/substrate/client/api/src/backend.rs#L339). Similar for `Client::storage_pairs()` and `Client::child_storage_keys()`.

dimartiro reviewed (changes requested):
LGTM but there are some things we can check or / and change before merging

haikoschol committed:
Merge branch 'development' into haiko/read-only-storage-state

haikoschol committed:
address some review feedback

haikoschol committed:
correctly set IterArgs.StartAtExclusive

timwu20 commented on internal/client/adapter/client_adapter.go:
```diff
@@ -276,39 +291,69 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) GetHighestRoundAndSetID() (uin
+// GetBlockHashesBySlot gets all block hashes that were produced in the given slot.
 func (ca *ClientAdapter[H, Hasher, N, E, Header]) GetBlockHashesBySlot(slotNum uint64) ([]common.Hash, error) {
-	panic("unimplemented")
+	children, err := ca.client.Children(ca.client.Info().FinalizedHash)
```
I don't think this method is called anywhere, so I think we can remove from `BlockState`.

timwu20 commented on internal/client/adapter/client_adapter.go:
```diff
@@ -398,8 +443,10 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) SetHighestRoundAndSetID(round,
+// GetRoundAndSetID returns the finalised round and setID
 func (ca *ClientAdapter[H, Hasher, N, E, Header]) GetRoundAndSetID() (uint64, uint64) {
-	panic("unimplemented")
+	round, setID, _ := ca.GetHighestRoundAndSetID()
```
don't think this is called anymore.

timwu20 commented on internal/client/adapter/client_adapter.go:
```diff
@@ -495,53 +542,208 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) StoreTrie(rtstorage.TrieState,
 }
 
 func (ca *ClientAdapter[H, Hasher, N, E, Header]) GenerateTrieProof(stateRoot common.Hash, keys [][]byte) (
 	[][]byte, error) {
```
any reason why this is still unimplemented?

haikoschol replied:
Originally the answer to that question was because I overlooked it. Now the answer is because I don't know how. 😅 Looking at the old implementation hasn't helped so far.

I also overlooked `StorageState.TrieState()`, which I've implemented [here](https://github.com/ChainSafe/gossamer/pull/4769/commits/e6f4a0e52aad7a37057f79277a1d985ad5b8180a).

dimartiro replied:
I think you can use [NewMerkleProof](https://github.com/ChainSafe/gossamer/blob/3e2c0a005dc914192014f4e8d0f34b6a6650f24b/pkg/trie/triedb/proof/proof.go#L20)

haikoschol replied:
@dimartiro thanks for pointing me towards `NewMerkleProof()`. There's just one issue; it expects a `hashdb.HashDB` and all i have is a `ClientAdapterDB` and a `statemachine.Backend`.

timwu20 commented on internal/client/adapter/client_adapter.go:
```diff
@@ -495,53 +542,208 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) StoreTrie(rtstorage.TrieState,
+
+	targetRoot := (*new(Hasher)).NewHash(root.ToBytes())
+
+	for i := 0; i < maxSearchDepth; i++ {
```
why do we need to search for this? We are searching based on the root hash?

haikoschol replied:
This is used in `GetStorage()`, which expects a state root hash instead of a block header hash. Is there another way to get the block hash than this brute-force search?

timwu20 commented on internal/client/adapter/client_adapter.go:
```diff
@@ -556,6 +758,138 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) SetBlockTree(blocktree *blockt
 	panic("unimplemented")
 }
 
+// StorageProvider impl
```
why does the `ClientAdapter` need to implement `StorageProvider`?  The intention was to implement it for the `Client`, and use methods from `StorageProvider` to implement these `StorageState` methods.

haikoschol replied:
Yeah, that makes sense. <strike>The issue description just says "Introduce `StorageProvider` interface and implement it" and since I was already implementing stuff on `ClientAdapter` I added it there without thinking it through.</strike>

edit: The issue description actually says "`GetStorageXXX` methods are easily accessible by translating and implementing the [`StorageProvider`](https://github.com/paritytech/polkadot-sdk/blob/030cb4a71b0b390626a586bfe7117b7c66b4700c/substrate/client/api/src/backend.rs#L418) trait for `Client`". Apparently I didn't read that properly.

I've moved the implementation to `Client` [here](https://github.com/ChainSafe/gossamer/commit/ec5f7d05b35e2d969138dcc6cf8aa8552a2495e1).

timwu20 commented on internal/client/api/backend.go:
```diff
@@ -352,3 +353,51 @@ type Backend[
+
+	// StorageKeys returns a [statemachine.KeysIter] that iterates over matching storage keys in that block
+	// given a blocks hash and a key prefix.
+	StorageKeys(hash H, prefix, startKey statemachine.StorageKey) (statemachine.KeysIter[H, Hasher], error)
```
this should return an instance of the local `KeysIter`.  You need to create the constructor for `KeysIter`  which calls  `backend.RawIter` with a supplied `IterArgs` ([link to code](https://github.com/paritytech/polkadot-sdk/blob/030cb4a71b0b390626a586bfe7117b7c66b4700c/substrate/client/api/src/backend.rs#L330)).

haikoschol replied:
Like [this](https://github.com/ChainSafe/gossamer/commit/37dd38d6340348582553d65235baaba0318e6c9a)?

haikoschol replied:
Based on [this comment](https://github.com/ChainSafe/gossamer/pull/4769#discussion_r2167582076) I've used `statemachine.KeysIter` and `statemachine.PairsIter` in `StorageProvider` and added the constructors to the `statemachine` package ([commit](https://github.com/ChainSafe/gossamer/pull/4769/commits/65615ce006a2865c7f63cf0adbb8fecf8af2d87c)).

timwu20 commented on internal/client/api/backend.go:
```diff
@@ -352,3 +353,51 @@ type Backend[
+// StorageProvider provides access to storage primitives
+type StorageProvider[H runtime.Hash, Hasher runtime.Hasher[H]] interface {
+	// Storage returns the value under the key in that block, given a blocks hash and a key.
+	Storage(hash H, key statemachine.StorageKey) (statemachine.StorageValue, error)
```
please use the correct primitives during this development phase.  We can consolidate the primitives after if it really doesn't make sense to differentiate between `storage.StorageData` and `statemachine.StorageValue`.

haikoschol committed:
Merge branch 'development' into haiko/read-only-storage-state

haikoschol committed:
use updated TrieLayout types/API

haikoschol committed:
use storage.StorageKey instead of statemachine.StorageKey in StorageProvider

haikoschol committed:
use local KeysIter/PairsIter in StorageProvider

haikoschol committed:
remove unused methods from BlockState and ClientAdapter

haikoschol committed:
move StorageProvider impl from ClientAdapter to Client

haikoschol committed:
Merge branch 'development' into haiko/read-only-storage-state

haikoschol committed:
remove another unused method from BlockState

haikoschol committed:
implement ClientAdapter.Leaves()

haikoschol committed:
implement ClientAdapter.TrieState()

haikoschol committed:
Revert "remove another unused method from BlockState"

This reverts commit d05af3e4b3603051e05c11f14fa027a35300a5df.

haikoschol committed:
maybe fix mock

haikoschol committed:
more mock fixery

haikoschol committed:
update a bazillion BlockState mocks

dimartiro commented on internal/client/adapter/client_adapter.go:
```diff
@@ -472,10 +491,6 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) SetHighestRoundAndSetID(round,
-	panic("unimplemented")
-}
-
 func (ca *ClientAdapter[H, Hasher, N, E, Header]) GetRuntime(blockHash common.Hash) (instance rt.Instance, err error) {
```
I think we should implement this in some point if we want to achieve a full read-only shim type. Do we have an issue for this one?

haikoschol replied:
Is it part of #4807?

dimartiro commented on internal/client/api/backend.go:
```diff
@@ -350,3 +352,226 @@ type Backend[
 }
+
+// KeysIter is an iterator over storage keys.
+type KeysIter[H runtime.Hash, Hasher runtime.Hasher[H]] struct {
```
We already have this defined in `internal/primitives/state-machine/backend.go` right? do we need it twice?

haikoschol replied:
That's how I understood "local `KeysIter`" in Tims comment [here](https://github.com/ChainSafe/gossamer/pull/4769#discussion_r2157985672). But I can change it to use the one in `statemachine` and move the constructors to that package.

dimartiro replied:
I think it's better to reuse the current one if we can

haikoschol replied:
[done](https://github.com/ChainSafe/gossamer/pull/4769/commits/65615ce006a2865c7f63cf0adbb8fecf8af2d87c)

dimartiro commented on internal/client/client_test.go:
```diff
@@ -63,6 +63,7 @@ var (
 	]] = &TestClient{}
+	_ api.StorageProvider[hash.H256, runtime.BlakeTwo256] = &TestClient{}
 )
 
```
I think you can add more tests cases here since you added the `api.StorageProvider` implementation for `Client`

haikoschol committed:
WIP GenerateTrieProof()

haikoschol committed:
use statemachine.KeysIter in StorageProvider

haikoschol committed:
fix(state): Change StorageState methods to always expect a block hash (#4792)

haikoschol committed:
update StorageState impl on ClientAdapter

After #4792, StorageState methods that expected a state root hash now
take a block hash. This updates all affected StorageState methods on
ClientAdapter and simplifies their implementation.

haikoschol committed:
Merge branch 'development' into haiko/read-only-storage-state

haikoschol committed:
remove duplicate StorageProvider definition
